### PR TITLE
Update `Text.split` to take a `Vector Text` parameter

### PR DESCRIPTION
**Pull request by GregoryTravis**
### Pull Request Description

Allows you to pass a vector of delimiters to `split`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] The documentation has been updated, if necessary.
- [x] Screenshots/screencasts have been attached, if there are any visual changes. For interactive or animated visual changes, a screencast is preferred.
- [x] All code follows the
      [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md),
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
      and
      [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md)
      style guides. In case you are using a language not listed above, follow the [Rust](https://github.com/enso-org/enso/blob/develop/docs/style-guide/rust.md) style guide.
- All code has been tested:
  - [x] Unit tests have been written where possible.
  - [x] If GUI codebase was changed, the GUI was tested when built using `./run ide build`.

### CODE REVIEW

File: std-bits/base/src/main/java/org/enso/base/Regex_Utils.java [hunk 1]
```diff
@@ -108,4 +108,8 @@ public static String sql_like_pattern_to_regex(String sql_pattern) {
 
     return result.toString();
   }
+
+  public static String regexQuote(String pattern) {
+    return pattern.replaceAll("[.*+?^${}()|\\[\\]]", "\\\\$0");
```
Comment (radeusgd) [inline std-bits/base/src/main/java/org/enso/base/Regex_Utils.java:113]: Why are so many `\\\\` needed? 😮

File: test/Tests/src/Data/Text_Spec.enso [hunk 2]
```diff
@@ -274,7 +274,29 @@ spec =
             utf_8_whitespace.split "\s+" use_regex=True . should_equal utf_8_whitespace_split
             'abc  def\tghi'.split '\\s+' use_regex=True . should_equal ["abc", "def", "ghi"]
 
-        Test.specify "exmples should be correct" <|
+        Test.specify 'should be able to split with a vector of strings' <|
+            'azbzczdzezfzg'.split ['b', 'zez'] . should_equal ['az', 'zczd', 'fzg']
+            'blah x 123'.split [' ', ' x ' , 'x'] . should_equal ['blah', '', '', '123']
+            'a1b2c3d4e5f6g7h8'.split ['c', '5'] . should_equal ['a1b2', '3d4e', 'f6g7h8']
```
Comment (radeusgd) [inline test/Tests/src/Data/Text_Spec.enso]: I think ideally split should work consistently with Unicode.

Could you add these tests?
```
'aśbs\u0301c'.split 'ś' . should_equal ['a', 'b', 'c']
'aśbs\u0301c'.split ['ś'] . should_equal ['a', 'b', 'c']
'aśbs\u0301c'.split 's\u0301' . should_equal ['a', 'b', 'c']
'aśbs\u0301c'.split ['s\u0301'] . should_equal ['a', 'b', 'c']
'aśbs\u0301cdef'.split ['ś', 'de'] . should_equal ['a', 'b', 'c', 'f']
```

File: std-bits/base/src/main/java/org/enso/base/Regex_Utils.java (hunk 1 above)
Comment (jdunkerley) [reply std-bits/base/src/main/java/org/enso/base/Regex_Utils.java:113]: otherwise the `\\` escapes the `$`

**Review by radeusgd — COMMENTED** [review]
I think splitting should work consistently regardless of if we provide it with a single argument or with a Vector of arguments.

Please add tests that verify that Unciode behaviour (in non-regex mode) is consistent in case I provide only one delimiter or a Vector of delimiters.

I suspect that since the current implementation relies on creating a Regex to do a multi-delimiter split even in non-Regex mode, that will not handle Unicode normalization properly.

I think we should deal with Unicode normalization correctly wherever we can, as we are trying to do through the language. Most importantly we should do it _consistently_. We can decide that `split` does not need to handle Unicode Normalization (although then probably `find` and `==` shouldn't too, and I wouldn't say it's the best idea), but we shouldn't make it switch behaviour depending on if a single or multiple delimiters are provided (switching behaviour if `use_regex=True` is ok since we should just explain that our Regex engine is not capable of handling Unicode normalization correctly).

Ideally, I think we should just switch to a non-Regex solution for `split` on a `Vector` with `use_regex=False`.

File: std-bits/base/src/main/java/org/enso/base/Regex_Utils.java (hunk 1 above)
Comment (radeusgd) [reply std-bits/base/src/main/java/org/enso/base/Regex_Utils.java:113]: Oh!

File: test/Tests/src/Data/Text_Spec.enso (hunk 2 above)
Comment (jdunkerley) [reply test/Tests/src/Data/Text_Spec.enso]: Yes - @radeusgd is correct. 
@GregoryTravis and I spoke yesterday about using the Regex based split as quicker to implement.
I think we will need to go to a slower approach as we discussed yesterday.

If the delimiter is a length 1 vector we should fall to the standard approach.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 3]
```diff
@@ -329,37 +328,50 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+Text.split : Text | Vector Text  -> Case_Sensitivity -> Boolean -> Vector Text | Illegal_Argument
+Text.split self delimiter="," case_sensitivity=Case_Sensitivity.Sensitive use_regex=False =
+    delimiter_is_empty = case delimiter of
+        _ : Text -> delimiter.is_empty
+        _ : Vector -> delimiter.is_empty || delimiter.any (.is_empty)
+        _ -> Error.throw (Illegal_Argument.Error "The delimiter must be a Text or Vector of Texts.")
+    delimiter_is_singleton_vector = case delimiter of
+        _ : Vector -> delimiter.length == 1
+        _ -> False
+    if delimiter_is_empty then Error.throw (Illegal_Argument.Error "The delimiter cannot be empty or contain empty entries.") else
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: nitpick but I'd reword this
```suggestion
    if delimiter_is_empty then Error.throw (Illegal_Argument.Error "The delimiter cannot be empty.") else
```

If I provide a Vector for `delimiter` I provide many _delimiters_, not a single _delimiter with some entries_. So IMO speaking about _entries of a delimiter_ is confusing.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 4]
```diff
@@ -329,37 +328,50 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
-                True -> parenthesize (delimiter.map parenthesize . join '|')
-                False -> parenthesize (delimiter.map (s-> Regex_Utils.regexQuote s) . map parenthesize . join '|')
-            self.split combined_delimiter case_sensitivity=case_sensitivity only_first=only_first use_regex=True
-        _ ->
-            msg = "Cann't split on " + delimiter.to_display_text
-            Error.throw (Illegal_Argument.Error msg)
+                True -> case delimiter of
+                    _ : Text ->
+                        Helpers.regex_assume_default_locale case_sensitivity <|
+                            case_insensitive = case_sensitivity.is_case_insensitive_in_memory
```
Comment (radeusgd) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: This can be simplified after #6150, right?

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 5]
```diff
@@ -323,26 +323,55 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+    delimiter_is_empty = case delimiter of
+        _ : Text -> delimiter.is_empty
+        _ : Vector -> delimiter.is_empty || delimiter.any (.is_empty)
+        _ -> Error.throw (Illegal_Argument.Error "The delimiter must be a Text or Vector of Texts.")
+    delimiter_is_singleton_vector = case delimiter of
+        _ : Vector -> delimiter.length == 1
+        _ -> False
+    if delimiter_is_empty then Error.throw (Illegal_Argument.Error "The delimiter cannot be empty or contain empty entries.") else
+        # If it's a vector of one element, just call it on that one element.
+        if delimiter_is_singleton_vector then self.split delimiter=(delimiter.at 0) case_sensitivity=case_sensitivity use_regex=use_regex else
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
        if delimiter_is_singleton_vector then self.split delimiter=delimiter.first case_sensitivity use_regex else
```
Prefer first to `.at 0`
Personal preference but no need to use named arguments here I don't think.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 6]
```diff
@@ -323,26 +323,55 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+                            delimiters.at i-1 . codeunit_end
+                        end = if i == delimiters.length then (Text_Utils.char_length self) else
+                            delimiters.at i . codeunit_start
+                        Text_Utils.substring self start end
+                True -> case delimiter of
+                    _ : Text ->
+                        Helpers.regex_assume_default_locale case_sensitivity <|
+                            case_insensitive = case_sensitivity.is_case_insensitive_in_memory
+                            compiled_pattern = Regex_2.compile delimiter case_insensitive=case_insensitive
+                            compiled_pattern.split self
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
                        case_insensitive = case_sensitivity.is_case_insensitive_in_memory
                        compiled_pattern = Regex_2.compile delimiter case_insensitive=case_insensitive
                        compiled_pattern.split self
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 7]
```diff
@@ -323,26 +323,55 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+                                    Text_Utils.span_of_all_multiple self delimiter
+                                Case_Sensitivity.Insensitive locale ->
+                                    Text_Utils.span_of_all_case_insensitive_multiple self delimiter locale.java_locale
+                            _ -> Error.throw (Illegal_Argument.Error "Cann't split on " + delimiter.to_display_text)
+                    Vector.new delimiters.length+1 i->
+                        start = if i == 0 then 0 else
+                            delimiters.at i-1 . codeunit_end
+                        end = if i == delimiters.length then (Text_Utils.char_length self) else
+                            delimiters.at i . codeunit_start
+                        Text_Utils.substring self start end
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: I wonder if having this as a private static method would help on readability. It's quite nested by this point.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 8]
```diff
@@ -323,26 +323,55 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+                                Case_Sensitivity.Sensitive ->
+                                    Text_Utils.span_of_all self delimiter
+                                Case_Sensitivity.Insensitive locale ->
+                                    Text_Utils.span_of_all_case_insensitive self delimiter locale.java_locale
+                            _ : Vector -> case case_sensitivity of
+                                Case_Sensitivity.Sensitive ->
+                                    Text_Utils.span_of_all_multiple self delimiter
+                                Case_Sensitivity.Insensitive locale ->
+                                    Text_Utils.span_of_all_case_insensitive_multiple self delimiter locale.java_locale
+                            _ -> Error.throw (Illegal_Argument.Error "Cann't split on " + delimiter.to_display_text)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
                            _ -> Error.throw (Illegal_Argument.Error "Cannot split on " + delimiter.to_display_text)
```

Comment (jdunkerley) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Logically this path is impossible.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 9]
```diff
@@ -323,26 +323,55 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+                            _ : Text -> case case_sensitivity of
+                                Case_Sensitivity.Sensitive ->
+                                    Text_Utils.span_of_all self delimiter
+                                Case_Sensitivity.Insensitive locale ->
+                                    Text_Utils.span_of_all_case_insensitive self delimiter locale.java_locale
+                            _ : Vector -> case case_sensitivity of
+                                Case_Sensitivity.Sensitive ->
+                                    Text_Utils.span_of_all_multiple self delimiter
+                                Case_Sensitivity.Insensitive locale ->
+                                    Text_Utils.span_of_all_case_insensitive_multiple self delimiter locale.java_locale
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: If we renamed `span_of_all_multiple` to `span_of_all` I think the polyglot invocation would mean we could do just:

```suggestion
                        case case_sensitivity of
                                Case_Sensitivity.Sensitive ->
                                    Text_Utils.span_of_all self delimiter
                                Case_Sensitivity.Insensitive locale ->
                                    Text_Utils.span_of_all_case_insensitive self delimiter locale.java_locale
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 10]
```diff
@@ -323,26 +323,55 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+                    _ : Text ->
+                        Helpers.regex_assume_default_locale case_sensitivity <|
+                            case_insensitive = case_sensitivity.is_case_insensitive_in_memory
+                            compiled_pattern = Regex_2.compile delimiter case_insensitive=case_insensitive
+                            compiled_pattern.split self
+                    _ : Vector ->
+                        parenthesize s = "(?:" + s + ")"
+                        combined_delimiter = parenthesize (delimiter.map parenthesize . join '|')
+                        self.split combined_delimiter case_sensitivity=case_sensitivity use_regex=True
+                    _ -> Error.throw (Illegal_Argument.Error "Cann't split on " + delimiter.to_display_text)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: as above - should be impossible to get here.

File: test/Tests/src/Data/Text_Spec.enso [hunk 11]
```diff
@@ -18,6 +18,8 @@ from Standard.Test import Test, Test_Suite
 import Standard.Test.Extensions
 import Standard.Base.Data.Text.Extensions
 
+polyglot java import java.lang.IllegalArgumentException
+
```
Comment (jdunkerley) [inline test/Tests/src/Data/Text_Spec.enso]: This shouldn't be need in the test scope - we should be handling before getting here.

File: std-bits/base/src/main/java/org/enso/base/Text_Utils.java [hunk 12]
```diff
@@ -355,6 +357,53 @@ public static List<Utf16Span> span_of_all(String haystack, String needle) {
...
+   * Find spans of all occurrences of a set of needles within the haystack.
+   *
+   * @param haystack the string to search
+   * @param needles the substrings that are searched for
+   * @return a list of UTF-16 code unit spans at which the needle occurs in the haystack
+   */
+  public static List<Utf16Span> span_of_all_multiple(String haystack, List<String> needles) {
+    if (needles.isEmpty())
+      throw new IllegalArgumentException(
+              "The operation `span_of_all_multiple` does not support searching for an empty term.");
```
Comment (jdunkerley) [inline std-bits/base/src/main/java/org/enso/base/Text_Utils.java]: I'd be tempted to do the individual needle check here as well `needles.stream().anyMatch(String::isEmpty)`.

File: std-bits/base/src/main/java/org/enso/base/Text_Utils.java [hunk 13]
```diff
@@ -355,6 +357,53 @@ public static List<Utf16Span> span_of_all(String haystack, String needle) {
...
+    for (String needle : needles) {
+        if (needle.isEmpty())
+          throw new IllegalArgumentException(
+                  "The operation `span_of_all_multiple` does not support searching for an empty term.");
+    }
+    if (haystack.isEmpty()) return List.of();
+
+    List<StringSearch> stringSearches = IntStream.range(0, needles.size())
+            .mapToObj(i -> new StringSearch(needles.get(i), haystack))
+            .collect(Collectors.toList());
```
Comment (jdunkerley) [inline std-bits/base/src/main/java/org/enso/base/Text_Utils.java]: from a pure pedant makes things a little cleaner we could collect into an array here as then needs less `get` calls.

File: std-bits/base/src/main/java/org/enso/base/Text_Utils.java [hunk 14]
```diff
@@ -477,6 +526,61 @@ public static List<GraphemeSpan> span_of_all_case_insensitive(
...
+        // No more matches.
+        break;
+      }
+      int matchLength = stringSearches.get(earliestIndex).getMatchLength();
+      GraphemeSpan graphemeSpan = findExtendedSpan(foldedHaystack, earliestStart, matchLength);
+      occurrences.add(graphemeSpan);
+      ix = graphemeSpan.codeunit_end;
+    }
+    return occurrences;
+  }
```
Comment (jdunkerley) [inline std-bits/base/src/main/java/org/enso/base/Text_Utils.java]: From a DRY perspective:

```suggestion
  public static List<GraphemeSpan> span_of_all_case_insensitive_multiple(
          String haystack, List<String> needles, Locale locale) {
    try {
      CaseFoldedString foldedHaystack = CaseFoldedString.fold(haystack, locale);
      List<String> foldedNeedles = needles.stream()
              .map(needle -> CaseFoldedString.simpleFold(needle, locale))
              .collect(Collectors.toList());
      var foldedSpans = span_of_all(foldedHaystack.foldedString, foldedNeedles);

      List<GraphemeSpan> occurrences =
              foldedSpans.stream()
                      .map(span -> findExtendedSpan(foldedHaystack, span.start, span.length))
                      .collect(Collectors.toList());
      return occurrences;
    } catch (IllegalArgumentException e) {
      throw new IllegalArgumentException(
              "The operation `span_of_all_case_insensitive_multiple` does not support searching for an empty term.");
    }
  }
```

No guarantee that works as wrote in a text editor!

**Review by jdunkerley — CHANGES_REQUESTED** [review]

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 5 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: > Personal preference but no need to use named arguments here I don't think.

If we ever refactor the method and change what arguments are supported, it may be easier to notice if we forgot to amend this call-site if the arguments are named. If they are only re-ordered, it will still work.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 7 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: +1

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 9 above)
Comment (radeusgd) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Polyglot method overloading is very funky. And also IIRC it adds more overhead to the calls if there are overloads (but that would be probably a question to @JaroslavTulach). Anyway I know he advised me to not rely on Polyglot overload calls if I don't have to. So I think it's better to keep the method variants have different names.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 5 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Yes, I lean towards keeping it unambiguous even if it's wordy.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 8 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Yes, but only because it's checked earlier; a refactoring might change that.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 9 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: I agree with Radek here; it would work but it's a bit of magic and it wouldn't be obvious that it was happening without careful reading.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 4 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Yes, this is removed in that PR.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso (hunk 10 above)
Comment (GregoryTravis) [reply distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: Same -- I'd like to leave it in, especially now that it's in a separate function. I assume there's no runtime penalty for a case that's unlikely to ever be reached.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 15]
```diff
@@ -1334,3 +1350,22 @@ slice_text text char_ranges =
...
        sb.append text char_range.start char_range.end
     sb.toString
+
+## PRIVATE
+
+   Find occurrences of delimiters in a string.
+split_find_delimiters : Text -> Text | Vector Text  -> Case_Sensitivity -> Vector Text | Illegal_Argument
+split_find_delimiters input delimiter case_sensitivity =
+    Vector.from_polyglot_array <|
+    case delimiter of
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
    Vector.from_polyglot_array <| case delimiter of
```

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 16]
```diff
@@ -1334,3 +1350,22 @@ slice_text text char_ranges =
...
+            Case_Sensitivity.Sensitive ->
+                Text_Utils.span_of_all input delimiter
+            Case_Sensitivity.Insensitive locale ->
+                Text_Utils.span_of_all_case_insensitive input delimiter locale.java_locale
+        _ : Vector -> case case_sensitivity of
+            Case_Sensitivity.Sensitive ->
+                Text_Utils.span_of_all_multiple input delimiter
+            Case_Sensitivity.Insensitive locale ->
+                Text_Utils.span_of_all_case_insensitive_multiple input delimiter locale.java_locale
+        _ -> Error.throw (Illegal_Argument.Error "Can't split on " + delimiter.to_display_text)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: don't think we need this line as `delimiter_is_empty` already checks this.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 17]
```diff
@@ -323,26 +323,42 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+                True -> case delimiter of
+                    _ : Text ->
+                        case_insensitive = case_sensitivity.is_case_insensitive_in_memory
+                        compiled_pattern = Regex_2.compile delimiter case_insensitive=case_insensitive
+                        compiled_pattern.split self
+                    _ : Vector ->
+                        parenthesize s = "(?:" + s + ")"
+                        combined_delimiter = parenthesize (delimiter.map parenthesize . join '|')
+                        self.split combined_delimiter case_sensitivity=case_sensitivity use_regex=True
+                    _ -> Error.throw (Illegal_Argument.Error "Can't split on " + delimiter.to_display_text)
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]: don't think we need this line as `delimiter_is_empty` already checks this.

File: distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso [hunk 18]
```diff
@@ -323,26 +323,42 @@ Text.match self pattern=".*" case_sensitivity=Case_Sensitivity.Sensitive =
...
+Text.split self delimiter="," case_sensitivity=Case_Sensitivity.Sensitive use_regex=False =
+    delimiter_is_empty = case delimiter of
+        _ : Text -> delimiter.is_empty
+        _ : Vector -> delimiter.is_empty || delimiter.any (.is_empty)
+        _ -> Error.throw (Illegal_Argument.Error "The delimiter must be a Text or Vector of Texts.")
+    delimiter_is_singleton_vector = case delimiter of
+        _ : Vector -> delimiter.length == 1
+        _ -> False
+    if delimiter_is_empty then Error.throw (Illegal_Argument.Error "The delimiter cannot be empty.") else
+        # If it's a vector of one element, just call it on that one element.
```
Comment (jdunkerley) [inline distribution/lib/Standard/Base/0.0.0-dev/src/Data/Text/Extensions.enso]:
```suggestion
    if delimiter_is_empty then Error.throw (Illegal_Argument.Error "The delimiter cannot be empty.") else
        # If it's a vector of one element, just call it on that one element.
        delimiter_is_singleton_vector = case delimiter of
            _ : Vector -> delimiter.length == 1
            _ -> False
```

**Review by jdunkerley — APPROVED** [review]
A couple of tiny nit pics but looks good

File: test/Tests/src/Data/Text_Spec.enso [hunk 19]
```diff
@@ -274,7 +274,55 @@ spec =
...
+            'aśbs\u0301c'.split 's\u0301' . should_equal ['a', 'b', 'c']
+            'aśbs\u0301c'.split ['s\u0301'] . should_equal ['a', 'b', 'c']
+            'aśbs\u0301cdef'.split ['ś', 'de'] . should_equal ['a', 'b', 'c', 'f']
+
+        Test.specify "should handle unicode normalization the same for single and multiple delimiters, case-insensitively" <|
+            'aśbs\u0301c'.split 'ś' case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
+            'aśbs\u0301c'.split ['ś'] case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
+            'aśbs\u0301c'.split 's\u0301' case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
+            'aśbs\u0301c'.split ['s\u0301'] case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
+            'aśbs\u0301cdef'.split ['ś', 'de'] case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c', 'f']
```
Comment (radeusgd) [inline test/Tests/src/Data/Text_Spec.enso]:
```suggestion
            'aśbS\u0301c'.split 'ś' case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
            'aśbS\u0301c'.split ['ś'] case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
            'aśbS\u0301c'.split 's\u0301' case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
            'aśbS\u0301c'.split ['s\u0301'] case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c']
            'aśbS\u0301cdef'.split ['ś', 'de'] case_sensitivity=Case_Sensitivity.Insensitive . should_equal ['a', 'b', 'c', 'f']
```

This test wasn't really checking that the thing works case insensitiviely, maybe worth swapping case of some of these `Ś`s.

File: test/Tests/src/Data/Text_Spec.enso [hunk 20]
```diff
@@ -274,7 +274,55 @@ spec =
             utf_8_whitespace.split "\s+" use_regex=True . should_equal utf_8_whitespace_split
             'abc  def\tghi'.split '\\s+' use_regex=True . should_equal ["abc", "def", "ghi"]
 
-        Test.specify "exmples should be correct" <|
+        Test.specify 'should be able to split with a vector of strings' <|
+            'azbzczdzezfzg'.split ['b', 'zez'] . should_equal ['az', 'zczd', 'fzg']
+            'blah x 123'.split [' ', ' x ' , 'x'] . should_equal ['blah', '', '', '123']
```
Comment (radeusgd) [inline test/Tests/src/Data/Text_Spec.enso]: Hmm, I just noticed this case and it is interesting. Maybe we should specify what happens when multiple delimiters overlap?


My intuition is it may be best to yield `['blah', '123']` here - i.e. take the 'largest' match. But I'm not sure how much that complicates stuff - so not suggesting to necessarily implement it - just wondering what is a good approach.

I'd consider adding one more test like this:
```
'abcdef'.split ['bc', 'cd'] == ???
```

I wonder what the answer will be? I guess probably `['a', 'def']`? Or is it `['a', 'ef']`? Both seem to make some sense (first that it matches the first delimiter, second that it does a 'union of all matched delimiter sections').

I'd appreciate adding this test, just to sanity check that we are not getting anything unexpected here (for example, I'd not expect to have any `''` in the result here).

Comment (jdunkerley) [reply test/Tests/src/Data/Text_Spec.enso]: The handling as Greg has is what I specified with him. 

image

It is what happens if you put the sequence into a regex split (i.e. split on `( | x |x)`. It is a bit of a degenerate case but is reasonable response.

Comment (radeusgd) [reply test/Tests/src/Data/Text_Spec.enso]: Fair. Still, I'd appreciate the `'abcdef'.split ['bc', 'cd']` test just to 'document' the behaviour, if we may.

Comment (GregoryTravis) [reply test/Tests/src/Data/Text_Spec.enso]: Added more test cases.
